packages/pds/src/api/com/atproto/sso/getCallback.ts

```diff
@@ -0,0 +1,555 @@
+import {
+  AuthRequiredError,
+  InternalServerError,
+  InvalidRequestError,
+} from "@atproto/xrpc-server";
+import { AppContext } from "../../../../context";
+import { Server } from "../../../../lexicon";
+import * as cookie from "cookie";
+import { base64url } from "jose";
+import {
+  UnauthorizedClientError,
+} from "@atproto/oauth-provider";
+import { DidDocument } from "@atproto/common";
+import { validateInputsForLocalPds } from "../server/createAccount";
+import { didDocForSession, safeResolveDidDoc } from "../server/util";
+import {
+  AccountStatus,
+  formatAccountStatus,
+} from "../../../../account-manager/account-manager";
+import { syncEvtDataFromCommit } from "../../../../sequencer";
+import { NullOutput, UserServiceAuthOutput } from "../../../../auth-verifier";
+import {
+  InputSchema,
+  OutputSchema,
+} from "../../../../lexicon/types/com/atproto/server/createAccount";
+import { INVALID_HANDLE } from "@atproto/syntax";
+import { softDeleted } from "../../../../db";
+import { resultPassthru } from "../../../proxy";
+import { GitHubEmailsSchema, OauthResponse, oauthResponseSchema, OidcClaims, oidcClaimsSchema } from "../../../../sso/db/schema/identity-provider";
+import { ActorAccount } from "../../../../account-manager/helpers/account";
+import { ssoLogger as log } from "../../../../logger";
+
+export default function (server: Server, ctx: AppContext) {
+  server.com.atproto.sso.getCallback({
+    auth: ctx.authVerifier.userServiceAuthOptional,
+    handler: async ({ params, auth, req }) => {
+      if (ctx.entrywayAgent) {
+        return resultPassthru(
+          await ctx.entrywayAgent.com.atproto.sso.getCallback(
+            params,
+            ctx.entrywayPassthruHeaders(req),
+          ),
+        )
+      }
+
+      if (params.error) {
+        const { error, error_description } = params;
+        const description = error_description && ` (${error_description})`;
+
+        log.error(`Callback response returned error: ${error}${description}`);
+
+        throw new InvalidRequestError(`Callback response returned error: ${error}${description}`);
+      }
+
+      if (!params.code) {
+        log.error(params, "Missing code in callback response");
+
+        throw new InvalidRequestError(`Missing code in callback response`);
+      }
+
+      const { code, state } = params;
+
+      if (!req.headers.cookie) {
+        throw new InvalidRequestError("Missing cookie header");
+      }
+
+      let callbackId: string | undefined = undefined;
+
+      if (typeof req.headers.cookie === 'string') {
+        try {
+          const cookies: Record<string, string | undefined>
+            = cookie.parse(req.headers.cookie);
+
+          callbackId = cookies["atproto-callback"];
+        } catch (err) {
+          throw new InvalidRequestError(`Invalid cookie header: ${err}`);
+        }
+      }
+
+      if (!callbackId) {
+        throw new InvalidRequestError("Missing callback cookie");
+      }
```

Comment:
You don't really need the cookie here, since you have that from the `state` parameter.

Reply:
I rewrote the cookie logic to do encryption, so that the state inside the cookie can provide CSRF protection.